proposals/0000-import-shadowing.rst

```diff
@@ -0,0 +1,362 @@
+
+ zip = ...
+
+to just
+
+::
+
+ module Mod where
+
+ zip = ...
+
+The above example is taken directly from `the "Import" page of the
+Haskell Wiki <https://wiki.haskell.org/Import>`_.
+
+Effect and Interactions
+-----------------------
+* Beside intra-module references, the other place where top-level
+  bindings can be used is export specifications. It feels natural to
+  resolve exports in the same scope used for the module. For example, if
+  we have something like
+
+  ::
+
+   module A (foo) where
+
+   import B -- This exports "foo"
+
+   foo = ...
+
+  then the ``foo`` exported by ``A`` should be the one defined in
+  ``A``'s top-level.
+
+* When modules are reexported wholesale, shadowing doesn't come into
+  play and the original module's contents are exported:
+
+  ::
+
+   module A (module B) where
+
+   import B -- this exports "foo"
+
+   foo = ...
+
+  Here, it is ``B.foo`` that is (re-)exported by ``A``, not ``A.foo``.
+
+  If both ``module B`` and ``foo`` are exported, then that is the same
+  category of error as without this extension exporting ``module B``
+  and ``module C`` with conflicting names, and should be reported the
+  same way.
+   
+* Top-level bindings that shadow imported names should be regarding as
+  shadowing bindings for the purposes of ``-Wname-shadowing``.
+
+Costs and Drawbacks
+-------------------
+The usual drawback of language extensions leading to some language
+fragmentation.
+
+Users new to Haskell seem to find this idea intuitive. We have
+gathered decade+-long experience with a Haskell compiler that uses
+import shadowing (and doesn't even let users turn it off), with a
+Haskell code base of several million lines of code that sees work from
+both experienced Haskell developers as well as people with a
+non-software-engineering background whose introduction to Haskell was
+via this compiler. There's no record of either novices (learning only
+the import-shadowing behaviour) or experienced Haskellers (who are
+used to imports being in the same scope as top-level definitions) ever
+getting into trouble due to this difference to Haskell 2010.
+
+
+Backward Compatibility
+----------------------
+Haskell 2010 doesn't have a mechanism for shadowing imported names,
+and valid Haskell 2010 programs retain their exact meanings with
+``ImportShadowing`` turned on. The proposed extension only makes
+previously unaccepted programs accepted by the scope checker.
+
+So this is a "-1"-impact change: it doesn't break existing code, and
+"un-breaks" existing broken code.
+
+A second-order effect of this is that using this extension can lead to
+preemptive forward compatibility. Adding a new export to ``Prelude``
+can lead to breakage just by virtue of existing code defining and
+using top-level definitions with the same name. With
+``ImportShadowing``, the existing intra-module references keep their
+meaning and there is no migration needed to accomodate the new
+``Prelude`` names.
+
+Alternatives
+------------
```

Comment:
You could also mention @goldfirere later imports shadow earlier imports as an alternative.